tedana/stats.py

```diff
@@ -0,0 +1,149 @@
+"""
+Statistical functions
+"""
+import logging
+
+import numpy as np
+from scipy import stats
+
+from tedana import utils
+
+LGR = logging.getLogger(__name__)
+
+
+def getfbounds(n_echos):
+    """
+    Gets F-statistic boundaries based on number of echos
+
+    Parameters
+    ----------
+    n_echos : :obj:`int`
+        Number of echoes
+
+    Returns
+    -------
+    fmin, fmid, fmax : :obj:`float`
+        F-statistic thresholds for alphas of 0.05, 0.025, and 0.01,
+        respectively.
+    """
+    f05 = stats.f.ppf(q=(1 - 0.05), dfn=1, dfd=(n_echos - 1))
+    f025 = stats.f.ppf(q=(1 - 0.025), dfn=1, dfd=(n_echos - 1))
+    f01 = stats.f.ppf(q=(1 - 0.01), dfn=1, dfd=(n_echos - 1))
+    return f05, f025, f01
+
+
+def computefeats2(data, mmix, mask, normalize=True):
+    """
+    Converts `data` to component space using `mmix`
+
+    Parameters
+    ----------
+    data : (S x T) array_like
+        Input data
+    mmix : (T [x C]) array_like
+        Mixing matrix for converting input data to component space, where `C`
+        is components and `T` is the same as in `data`
+    mask : (S,) array_like
+        Boolean mask array
+    normalize : bool, optional
+        Whether to z-score output. Default: True
+
+    Returns
+    -------
+    data_Z : (S x C) :obj:`numpy.ndarray`
+        Data in component space
+    """
+    if data.ndim != 2:
+        raise ValueError('Parameter data should be 2d, not {0}d'.format(data.ndim))
+    elif mmix.ndim not in [2]:
+        raise ValueError('Parameter mmix should be 2d, not '
+                         '{0}d'.format(mmix.ndim))
+    elif mask.ndim != 1:
+        raise ValueError('Parameter mask should be 1d, not {0}d'.format(mask.ndim))
+    elif data.shape[0] != mask.shape[0]:
+        raise ValueError('First dimensions (number of samples) of data ({0}) '
+                         'and mask ({1}) do not match.'.format(data.shape[0],
+                                                               mask.shape[0]))
+    elif data.shape[1] != mmix.shape[0]:
+        raise ValueError('Second dimensions (number of volumes) of data ({0}) '
+                         'and mmix ({1}) do not match.'.format(data.shape[0],
+                                                               mmix.shape[0]))
+
+    # demean masked data
+    data_vn = stats.zscore(data[mask], axis=-1)
+
+    # get betas of `data`~`mmix` and limit to range [-0.999, 0.999]
+    data_R = get_coeffs(data_vn, mmix, mask=None)
+    data_R[data_R < -0.999] = -0.999
```

Comment:
This is somewhat pedantic, but here is an example of a "mysterious number." The comment above helps, but if for some reason we were to change those values, we would have to manually catch them and not make typos. Review helps for that, but I am biased towards less work for reviewers. ; - )

Reply:
I have concerns about the calculation of the beta-values and their conversion to z-values as implemented here (see #178 and #223). My hope is that we will fix those issues (and thus eliminate the need for this weird cropping), which will also involve removing these hard-coded values.

Reply:
Oh, I hope I was clear, I don't think that this particular case should stop us from merging, and I know there are discussions about it, but I wanted to give a more concrete example for my comment. In fact, I DON'T think this should be adjusted in the PR because that's independent to the refactor.